runtime/participant-runtime.ss

```diff
@@ -767,3 +774,354 @@
+
+
+;; ------------------ Libp2p client methods
+;;
+;; For a complete reference of libp2p API.
+;; See: https://github.com/vyzo/gerbil-libp2p#libp2p-api
+;;
+;; To understand what multiaddresses are,
+;; see: https://github.com/multiformats/multiaddr
+;;
+;; The implementation of client methods here are influenced by those found in:
+;; https://github.com/vyzo/gerbil-libp2p/blob/master/example/libp2p-chat.ss
+
+
+;; FIXME:
+;; In the upstream branch,
+;; the `start-libp2p-daemon!' procedure doesn't redirect output
+;; to a logfile, instead just directly outputs it to the terminal.
+;;
+;; To fix this we have to:
+;; 1. Upstream to gambit
+;;    Extend open-process to allow you to redirect to file or file descriptor.
+;;    (In our case this allows us to redirect the output / error messages to a log-file)
+;;    See how `run-program` in uiop does this:
+;;    https://common-lisp.net/project/asdf/uiop.html#UIOP_002fRUN_002dPROGRAM
+;;   
+;; 2. Upstream to gerbil-libp2p
+;;    process-options to use the redirecting option
+;;    for writing the error logs to a file instead
+;;    of the console.
+;;
+;; In the short run we use the shell to redirect error logs to a file (see the implementation below).
+(def (start-libp2p-daemon! host-addresses: (host-addrs #f) daemon: (bin "p2pd")
+                           options: (options [])
+                           address: (sock #f)
+                           wait: (timeo 0.4)
+                           p2pd-log-path: (p2pd-log-path (log-path "p2pd.log")))
+  (cond
+   ((current-libp2p-daemon)
+    => values)
+   (else
+    (let* ((path (or sock (string-append "/tmp/p2pd." (number->string (getpid)) ".sock")))
+           (addr (string-append "/unix" path))
+
+           (raw-cmd (escape-shell-tokens [bin "-q" "-listen" addr
+                                          (if host-addrs ["-hostAddrs" host-addrs] [])...
+                                          options ...]))
+           (cmd (format "{ echo ~a ; exec ~a ; } < /dev/null >> ~a 2>&1"
+                 raw-cmd raw-cmd p2pd-log-path))
+
+           (proc (open-process [path: "/bin/sh" arguments: ["-c" cmd]]))
+
+           (d (daemon proc path)))
+      (cond
+       ((process-status proc timeo #f)
+        => (lambda (status)
+             (error "p2pd exited prematurely" status))))
+      (current-libp2p-daemon d)
+      d))))
+
+;; NOTE: This is needed to call initialize our version
+;; of the daemon process (by `start-libp2p-daemon!'),
+;; see the implementation above.
+;; TODO: Once the above changes are upstreamed for `start-libp2p-daemon!',
+;; this command can be made obsolete too.
+(def (open-libp2p-client host-addresses: (host-addresses #f) options: (args [])  address: (sock #f)  wait: (timeo 12) (path #f)) ;; Extra arguments host-address and options
+  (let (d (start-libp2p-daemon! host-addresses: host-addresses options: args address: sock wait: timeo)) ;; Should go with host-address/tranpsort/port
```

Comment:
These lines are very long and it makes it hard to read; maybe wrap them, with one key/value per line?